raw_linux.go

```diff
@@ -310,38 +281,84 @@ func (p *packetConn) handleStats(s unix.TpacketStats) *Stats {
 // sysSocket is the default socket implementation.  It makes use of
 // Linux-specific system calls to handle raw socket functionality.
 type sysSocket struct {
-	fd int
+	f  *os.File
+	rc syscall.RawConn
+}
+
+func (s *sysSocket) SetDeadline(t time.Time) error {
+	return s.f.SetDeadline(t)
+}
+
+func (s *sysSocket) SetReadDeadline(t time.Time) error {
+	return s.f.SetReadDeadline(t)
+}
+
+func (s *sysSocket) SetWriteDeadline(t time.Time) error {
+	return s.f.SetWriteDeadline(t)
+}
+
+func (s *sysSocket) Bind(sa unix.Sockaddr) error {
+	var err error
+	cerr := s.rc.Control(func(fd uintptr) {
+		err = unix.Bind(int(fd), sa)
+	})
+	if err != nil {
+		return err
+	}
+	return cerr
+}
+
+func (s *sysSocket) Close() error {
+	return s.f.Close()
 }
 
-// Method implementations simply invoke the syscall of the same name, but pass
-// the file descriptor stored in the sysSocket as the socket to use.
-func (s *sysSocket) Bind(sa unix.Sockaddr) error { return unix.Bind(s.fd, sa) }
-func (s *sysSocket) Close() error                { return unix.Close(s.fd) }
-func (s *sysSocket) FD() int                     { return s.fd }
 func (s *sysSocket) GetSockopt(level, name int, v unsafe.Pointer, l uintptr) error {
-	_, _, err := unix.Syscall6(unix.SYS_GETSOCKOPT, uintptr(s.fd), uintptr(level), uintptr(name), uintptr(v), uintptr(unsafe.Pointer(&l)), 0)
-	if err != 0 {
-		return unix.Errno(err)
+	var err error
+	cerr := s.rc.Control(func(fd uintptr) {
+		_, _, errno := unix.Syscall6(unix.SYS_GETSOCKOPT, fd, uintptr(level), uintptr(name), uintptr(v), uintptr(unsafe.Pointer(&l)), 0)
+		if errno != 0 {
+			err = os.NewSyscallError("getsockopt", errno)
+		}
+	})
+	if err != nil {
+		return err
 	}
-	return nil
+	return cerr
 }
-func (s *sysSocket) Recvfrom(p []byte, flags int) (int, unix.Sockaddr, error) {
-	return unix.Recvfrom(s.fd, p, flags)
+
+func (s *sysSocket) Recvfrom(p []byte, flags int) (n int, addr unix.Sockaddr, err error) {
+	cerr := s.rc.Read(func(fd uintptr) bool {
+		n, addr, err = unix.Recvfrom(int(fd), p, flags)
+		return err != unix.EAGAIN
```

Comment:
Leave a small comment along the lines of: https://github.com/mdlayher/netlink/blob/master/conn_linux.go#L518-L521.